From ff491171eb5564f18e8e6efebb13996a9dea2fa7 Mon Sep 17 00:00:00 2001
From: Nirmoy Das <nirmoyd@nvidia.com>
Date: Wed, 10 Sep 2025 08:23:52 -0700
Subject: [PATCH] NVIDIA: SAUCE: iommu/arm-smmu-v3: Set DGX Spark iGPU default
 domain type to DMA

BugLink: https://bugs.launchpad.net/bugs/2129776

Force DGX Spark systems to use DMA translation as current drivers
require this.

Suggested-by: Jason Gunthorpe <jgg@nvidia.com>
Signed-off-by: Nirmoy Das <nirmoyd@nvidia.com>
Acked-by: Matthew R. Ochs <mochs@nvidia.com>
Acked-by: Carol L Soto <csoto@nvidia.com>
Acked-by: Abdur Rahman <abdur.rahman@canonical.com>
Acked-by: Jacob Martin <jacob.martin@canonical.com>
Signed-off-by: Ian May <ianm@nvidia.com>

(cherry picked from commit 54575cfd0fa45463ba90a504e01fcd27d9dd020f noble:linux-nvidia-6.17)
Signed-off-by: Jacob Martin <jacob.martin@canonical.com>
---
 drivers/iommu/arm/arm-smmu-v3/arm-smmu-v3.c | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/drivers/iommu/arm/arm-smmu-v3/arm-smmu-v3.c b/drivers/iommu/arm/arm-smmu-v3/arm-smmu-v3.c
index 5f933d806a..2d6bd7994e 100644
--- a/drivers/iommu/arm/arm-smmu-v3/arm-smmu-v3.c
+++ b/drivers/iommu/arm/arm-smmu-v3/arm-smmu-v3.c
@@ -4298,6 +4298,9 @@ static int arm_smmu_def_domain_type(struct device *dev)
 
 		if (IS_HISI_PTT_DEVICE(pdev))
 			return IOMMU_DOMAIN_IDENTITY;
+
+		if (pdev->vendor == PCI_VENDOR_ID_NVIDIA && pdev->device == 0x2E12)
+			return IOMMU_DOMAIN_DMA;
 	}
 
 	return 0;
From 00f46def1d5cdafdae7e7e864096a9b45c8255a7 Mon Sep 17 00:00:00 2001
From: Nirmoy Das <nirmoyd@nvidia.com>
Date: Thu, 20 Nov 2025 08:10:15 -0800
Subject: [PATCH] NVIDIA: SAUCE: iommu/arm-smmu-v3: Add two more DGX Spark iGPU
 IDs for existing iommu quirk

BugLink: https://bugs.launchpad.net/bugs/2132033

Add two more device IDs for the existing Spark iommu quirk.

Link: https://bugs.launchpad.net/ubuntu/+source/linux-nvidia-6.14/+bug/2132033

Signed-off-by: Nirmoy Das <nirmoyd@nvidia.com>
Acked-by: Jamie Nguyen <jamien@nvidia.com>
Acked-by: Carol L Soto <csoto@nvidia.com>
Acked-by: Abdur Rahman <abdur.rahman@canonical.com>
Acked-by: Noah Wager <noah.wager@canonical.com>
Signed-off--by: Brad Figg <bfigg@nvidia.com>

(cherry picked from commit ba9315458447d2677fb83131d78e1520f51b6c8a noble:linux-nvidia-6.17)
Signed-off-by: Jacob Martin <jacob.martin@canonical.com>
---
 drivers/iommu/arm/arm-smmu-v3/arm-smmu-v3.c | 4 +++-
 1 file changed, 3 insertions(+), 1 deletion(-)

diff --git a/drivers/iommu/arm/arm-smmu-v3/arm-smmu-v3.c b/drivers/iommu/arm/arm-smmu-v3/arm-smmu-v3.c
index 2d6bd7994e..07200fdb85 100644
--- a/drivers/iommu/arm/arm-smmu-v3/arm-smmu-v3.c
+++ b/drivers/iommu/arm/arm-smmu-v3/arm-smmu-v3.c
@@ -4299,7 +4299,9 @@ static int arm_smmu_def_domain_type(struct device *dev)
 		if (IS_HISI_PTT_DEVICE(pdev))
 			return IOMMU_DOMAIN_IDENTITY;
 
-		if (pdev->vendor == PCI_VENDOR_ID_NVIDIA && pdev->device == 0x2E12)
+		if (pdev->vendor == PCI_VENDOR_ID_NVIDIA &&
+		    (pdev->device == 0x2E12 || pdev->device == 0x2E2A ||
+		     pdev->device == 0x2E2B))
 			return IOMMU_DOMAIN_DMA;
 	}
 
From e6efeafde42ec871f7f6d66fd3662236827965c3 Mon Sep 17 00:00:00 2001
From: Abhishek Sahu <abhsahu@nvidia.com>
Date: Mon, 27 Apr 2026 05:04:56 +0000
Subject: [PATCH] NVIDIA: SAUCE: iommu/arm-smmu-v3: Use device ID range for DGX
 Spark iGPU iommu quirk

BugLink: https://bugs.launchpad.net/bugs/2150487

Replace the explicit DGX Spark iGPU device ID list with a range check
covering 0x2E00-0x2E3F to accommodate all possible DGX Spark iGPU PCI
device IDs without requiring individual additions.

The original quirk was introduced in
commit ab858638d96a ("NVIDIA: SAUCE: iommu/arm-smmu-v3:
Set DGX Spark iGPU default domain type to DMA") and
extended with two more IDs in
commit 8dc61abaa2eb ("NVIDIA: SAUCE: iommu/arm-smmu-v3:
Add two more DGX Spark iGPU IDs for existing iommu quirk").
Using a range avoids further per-ID additions as new DGX Spark
variants are introduced.

Signed-off-by: Abhishek Sahu <abhsahu@nvidia.com>
Acked-by: Jamie Nguyen <jamien@nvidia.com>
Acked-by: Matthew R. Ochs <mochs@nvidia.com>
Acked-by: Nirmoy Das <nirmoyd@nvidia.com>
Signed-off-by: Brad Figg <bfigg@nvidia.com>
---
 drivers/iommu/arm/arm-smmu-v3/arm-smmu-v3.c | 3 +--
 1 file changed, 1 insertion(+), 2 deletions(-)

diff --git a/drivers/iommu/arm/arm-smmu-v3/arm-smmu-v3.c b/drivers/iommu/arm/arm-smmu-v3/arm-smmu-v3.c
index 07200fdb85..4d34b8964a 100644
--- a/drivers/iommu/arm/arm-smmu-v3/arm-smmu-v3.c
+++ b/drivers/iommu/arm/arm-smmu-v3/arm-smmu-v3.c
@@ -4300,8 +4300,7 @@ static int arm_smmu_def_domain_type(struct device *dev)
 			return IOMMU_DOMAIN_IDENTITY;
 
 		if (pdev->vendor == PCI_VENDOR_ID_NVIDIA &&
-		    (pdev->device == 0x2E12 || pdev->device == 0x2E2A ||
-		     pdev->device == 0x2E2B))
+		    pdev->device >= 0x2E00 && pdev->device <= 0x2E3F)
 			return IOMMU_DOMAIN_DMA;
 	}
 
